feat: 사용자가 특정 API endpoint에 접근할 수 있는 권한이 있는지 검사할 수 있어야 합니다.

HiiWee opened the pull request:
## 📄 Summary

- `AuthenticationInterceptor`를 이용해 `비로그인`, `구매자`, `판매자`에 대한 엔드포인트 접근을 관리합니다.
  - 어노테이션 기반으로 `Controller` 메소드에서 `GetMapping`, `PostMapping`과
    동일한 위치에 사용하면 인터셉터가 검증해줍니다.
  - 아무 어노테이션도 붙이지 않았다면 `구매자`, `판매자` 모두 접근 가능한 엔드포인트로 인식합니다.
  - 만약 비로그인 사용자도 접근하는걸 허용하려면 `WebConfig`에서 `엔드포인트를 제외`시켜주어야 합니다.


### 접근 권한 명시

- 기존의 `@BuyerOnly`, `@SellerOnly`는 단일 접근을 허용할때 사용할 수 있습니다.
- `@Roles`는 `Role Enum`을 활용해 허용할 권한 집합을 명시합니다.
- 아무런 어노테이션도 붙이지 않는다면 모든 사용자(비로그인 포함)의 접근을 허용합니다.


```java
// 구매자, 판매자 모두 접근 가능
@Roles({Role.BUYER, Role.SELLER})
public void controllerEndpoint() {}

// 구매자 접근 가능 (2가지 방법)
@Roles(Role.BUYER)
public void controllerEndpoint() {}

@BuyerOnly
public void controllerEndpoint() {]

// 판매자 접근 가능 (2가지 방법)
@Roles({Role.SELLER}) 
public void controllerEndpoint() {}

@SellerOnly
public void controllerEndpoint() {}

// 모두 접근 가능
public void controllerEndpoint() {}
```

## 🙋🏻 More

close #217

HiiWee committed:
feat: 권한 검증을 위한 인터셉터를 생성한다

HiiWee committed:
docs: 로그인 API 문서 테스트에 링크 주석 추가

HiiWee committed:
feat: 인터셉터를 이용해 비로그인, 구매자, 판매자에 대한 엔드포인트 접근을 검증한다.

- 어노테이션 기반으로 Controller 메소드에서 GetMapping, PostMapping과
  동일한 위치에 사용하면 인터셉터가 검증해줍니다.
- 아무것도 붙이지 않았다면 구매자, 판매자 모두 접근 가능한 엔드포인트로 인식합니다.
- 만약 비로그인 사용자도 접근하는걸 허용하려면 WebConfig에서 엔드포인트를 제외시켜주어야 합니다.

HiiWee committed:
Merge branch 'dev' into feat/207

minseok-oh commented on src/main/java/com/wootecam/luckyvickyauction/core/auction/entity/AuctionEntity.java:
```diff
@@ -39,9 +39,9 @@ public class AuctionEntity {
 
     @Builder
     private AuctionEntity(Long id, Long sellerId, String productName, long originPrice, long currentPrice,
-                         long originStock, long currentStock, long maximumPurchaseLimitCount, PricePolicy pricePolicy,
-                         Duration variationDuration, ZonedDateTime startedAt, ZonedDateTime finishedAt,
-                         boolean isShowStock) {
+                          long originStock, long currentStock, long maximumPurchaseLimitCount, PricePolicy pricePolicy,
+                          Duration variationDuration, ZonedDateTime startedAt, ZonedDateTime finishedAt,
+                          boolean isShowStock) {
```
항상 감사합니다.. ☺️

chhs2131 commented:
접근 권한을 정리하면 이런 느낌이군요~
- 비로그인: WebConfig에 EndPoint 명시
- 로그인: Default
- 구매자: BuyerOnly 명시
- 판매자: SellerOnly 명시
- 둘다 붙이면 아무도 못씀!

minseok-oh commented on src/main/java/com/wootecam/luckyvickyauction/core/member/controller/AuthenticationInterceptor.java:
```diff
@@ -0,0 +1,52 @@
+        if (isBothPossiblePath(handlerMethod)) {
+            return true;
+        }
+        authorize(handlerMethod, signInInfo);
```
그쵸.. 이게 되는게 맞죠.. 3시간이 사라졌네요..
그래도 docsGiven이 String으로 세션을 준다는 걸 알았으니 좋습니다~

HiiWee commented:
> 접근 권한을 정리하면 이런 느낌이군요~
> 
> * 비로그인: WebConfig에 EndPoint 명시
> * 로그인: Default
> * 구매자: BuyerOnly 명시
> * 판매자: SellerOnly 명시
> * 둘다 붙이면 아무도 못씀!

고마워요 스피드 웨건! PR에 명시해놓을게요~

minseok-oh reviewed (dismissed):
덕분에 코드가 깔끔해지네요! 감사합니다!

chhs2131 reviewed (dismissed):
아주 필요하던 기능이네요 좋습니다 bb

HiiWee commented:
라이브 리뷰 from: @yudonggeun , @chhs2131 

- 어노테이션 기반으로 사용할거면 통일했으면 좋겠다는 의견
- 현재 상황은 비로그인 상황이 접근 가능한 엔드포인트를 `WebConfig`에 `exclude endpoint`로 명시해주어야 함
- 이를 어노테이션 기반으로 통일하기 위해 `@Role({Buyer, Seller})`와 같은 어노테이션을 만들어 Interceptor에서 모두 처리하도록 변경하고자 함

-> 결과적으로 아무 어노테이션도 없다면 Public Path 명시가 되어있으면 해당 권한을 가진 사용자만 접근할 수 있음을 의미합니다.

예상 결과
```java
// 구매자, 판매자 모두 접근 가능
@Role({Buyer, Seller})
public void controllerEndpoint() {}

@Role(Buyer) // 구매자 접근 가능
public void controllerEndpoint() {}

@Role(Seller) // 판매자 접근 가능
public void controllerEndpoint() {}

// 모두 접근 가능
public void controllerEndpoint() {}
```

HiiWee committed:
Merge branch 'dev' into feat/207

HiiWee committed:
refactor: AuthenticationInterceptor에서 모든 사용자의 권한을 검증하도록 변경합니다.

- 기존의 BuyerOnly, SellerOnly는 단일 권한을 허용할 때 사용할 수 있습니다.
- 추가로 Roles를 사용하면 Role enum으로 접근을 허용할 권한들을 설정할 수 있습니다.
- 만약 아무 어노테이션도 붙지 않은 엔드포인트라면 모든 사용자(비로그인 사용자 포함)의 접근을 허용 합니다.

HiiWee commented:
### 접근 권한 명시

- 기존의 `@BuyerOnly`, `@SellerOnly`는 단일 접근을 허용할때 사용할 수 있습니다.
- `@Roles`는 `Role Enum`을 활용해 허용할 권한 집합을 명시합니다.
- 아무런 어노테이션도 붙이지 않는다면 모든 사용자(비로그인 포함)의 접근을 허용합니다.


```java
// 구매자, 판매자 모두 접근 가능
@Roles({Role.BUYER, Role.SELLER})
public void controllerEndpoint() {}

// 구매자 접근 가능 (2가지 방법)
@Roles(Role.BUYER)
public void controllerEndpoint() {}

@BuyerOnly
public void controllerEndpoint() {]

// 판매자 접근 가능 (2가지 방법)
@Roles({Role.SELLER}) 
public void controllerEndpoint() {}

@SellerOnly
public void controllerEndpoint() {}

// 모두 접근 가능
public void controllerEndpoint() {}
```

HiiWee committed:
fix(test): pk값 검증을 테스트 코드에서 제외합니다.

HiiWee committed:
Merge branch 'dev' into feat/207

chhs2131 reviewed (approved):
좋아요 좋아요~
여전히 BuyerOnly, SellerOnly 둘다 붙이면 안되니 주의주의~!
